juju/client/proxy/factory.py

```diff
@@ -0,0 +1,39 @@
+from abc import ABC, abstractmethod
+from juju.client.proxy.kubernetes.proxy import KubernetesProxy
+
+class ProxyTypeNotDefinedError(Exception):
+    pass
+
+class UnknownProxyTypeError(Exception):
+    pass
+
+class Factory(ABC):
+
+    @abstractmethod
+    def proxy_from_config(self, conf):
+        raise NotImplementedError()
+
+class JujuDataFactory(Factory):
```

Comment:
I'm not sure why this is called `JujuDataFactory` -- that implies it's a "factory for Juju data", but it's a factory for a Proxy, right? Also, the use of Factory here gives me the Java-shivers. :-) It looks like this could just be replaced with a function (which are first-class in Python), for example `proxy_from_config()` as an ordinary function. One of the give-aways is that it's a class with only one method, and it doesn't use `self`. Here's a well-known talk by a Python programmer about this (though arguably he goes a bit too far): https://www.youtube.com/watch?v=o9pEzgHorH0

Reply:
Was trying to go for duck typing buck fell short.